flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/BlockStatementSplitter.java

```diff
@@ -0,0 +1,417 @@
+            }
+        }
+
+        rewriteBlocks.put(visitor.context, visitor.rewriter.getText());
+        return rewriteBlocks;
+    }
+
+    private static void mergeBlocks(
+            Map<String, List<String>> mergeA, Map<String, List<String>> mergeB) {
+
+        for (Entry<String, List<String>> entry : mergeA.entrySet()) {
+            mergeB.merge(
+                    entry.getKey(),
+                    entry.getValue(),
+                    (aBlocks, bBlocks) -> {
+                        List<String> merge = new ArrayList<>(aBlocks.size() + bBlocks.size());
+                        merge.addAll(aBlocks);
+                        merge.addAll(bBlocks);
+                        return merge;
+                    });
+        }
+    }
+
+    private static class BlockStatementVisitor extends JavaParserBaseVisitor<Void> {
+
+        private final List<BlockStatementVisitor> children = new ArrayList<>();
+
+        private final List<ContextTextPair> extractedSingleBlocks = new ArrayList<>();
+
+        private final String context;
+
+        private final CommonTokenStream tokenStream;
+
+        private final TokenStreamRewriter rewriter;
+
+        private final String parameters;
+
+        private int counter = 0;
+
+        private int bscCounter = 0;
+
+        private BlockStatementVisitor(String code, String context, String parameters) {
+            this.tokenStream = new CommonTokenStream(new JavaLexer(CharStreams.fromString(code)));
+            this.rewriter = new TokenStreamRewriter(tokenStream);
+            this.context = context;
+            this.parameters = parameters;
+        }
+
+        private BlockStatementVisitor(
+                TokenStreamRewriter rewriter, String context, String parameters) {
+            this.tokenStream = null;
+            this.rewriter = rewriter;
+            this.context = context;
+            this.parameters = parameters;
+        }
+
+        @Override
+        public Void visitStatement(StatementContext ctx) {
+
+            if (ctx.getChildCount() == 0) {
+                return null;
+            }
+
+            if (ctx.WHILE() != null) {
+                for (StatementContext statement : ctx.statement()) {
+                    BlockStatementVisitor visitor =
+                            new BlockStatementVisitor(
+                                    rewriter, context + "_whileBody" + counter++, parameters);
+                    children.add(visitor);
+                    visitor.visitStatement(statement);
+                }
+            } else if (ctx.IF() != null) {
+                for (StatementContext sc : ctx.statement()) {
+                    BlockStatementVisitor visitor =
+                            new BlockStatementVisitor(
+                                    rewriter, context + "_ifBody" + counter++, parameters);
+                    children.add(visitor);
+                    visitor.visitStatement(sc);
+                }
+            } else {
```

Comment:
What about `ctx.ELSE() != null`?

Reply:
Hi @tsreaper 
In fact we don't need `ctx.ELSE()` branch at all. I will try to explain why.

The initial input statement for `BlockStatementSplitter` can be two things:

- a statement that is represented like this:

```
	if (a[2] == 0) {
		a[2] = 1;
	} else {
		a[2] = b[2];
	}
```

- a block, that will be represented as this:

```
{
	if (a[2] == 0) {
		a[2] = 1;
	} else {
		a[2] = b[2];
	}
}
```

Mind that block has additional brackets `{ }` at the beginning and at the end. 

When `BlockStatementSplitter` is called from `BlockStatementRewriter` it will always be called by passing the statement, so without brackets `{ }`. The statement will never start from `else`, it will start from IF or WHILE (this is what we detect for now) or it will be what I call a "single line statement", meaning operation on a field, for example `a[2] = 1;`.

Lets look at the example. The `BlockStatementRewriter` called `BlockStatementSplitter::visitStatement` with below statement:

```
if (a[0] == 0) {
	if (a[2] == 0) {
		a[2] = 1;
	} else {
		a[2] = b[2];
	}
} else {
	a[0] = b[0];
	a[1] = b[1];
}
```

For this, the condition `if (ctx.WHILE() != null)` will be true. 
In this branch we will iterate through every `ctx.statement()`
The first item would be:

```
{
	if (a[2] == 0) {
		a[2] = 1;
	} else {
		a[2] = b[2];
	}
}
```
So the `TRUE` branch of the first level IF/ELSE expression. We will create a new context to indicate that we will process an IF(possibly also ELSE) statement and we will create new `BlockStatementVisitor` and call visitStatement method to process that block, since it can contain more nested structures we would like to rewrite. Mind that now the input argument is a "block", it starts and ends with brackets `{ }`

The second item in the iteration routine will be this:
```
{
	a[0] = b[0];
	a[1] = b[1];
}
```
So the `FALSE` branch of the first level IF/ELSE expression. Same as for TRUE, we will create new context and `BlockStatementVisitor` to process that block since it can contain more nested structures we would like to rewrite. In this case the input also would be the "block" since it starts and ends with brackets `{ }`, same as for TRUE branch.

In both new Visitors, conditions `ctx.WHILE() != null` and `ctx.IF() != null` will be false since now we have a "block", it starts from `{`. So we will land at `if (ctx.block() != null) ` line ~224 of `BlockStatementSplitter.java`. Since this is a block now, we need to iterate over all its statements hence `for (BlockStatementContext bsc : ctx.block().blockStatement())`

For `TRUE` branch from the initial statement it will be only one item:
```
if (a[2] == 0) {
	a[2] = 1;
} else {
	a[2] = b[2];
}
```
We will create new Visitor and process this statement further.

For `FALSE` branch, from the initial statement it will contain two elements. Those will be "single line elements":
- a[0] = b[0];
- a[1] = b[1];

Those will be added to `extractedSingleBlocks` blocks and extracted to separate method.

I hope that clear the logic a little bit.